### PR TITLE
feat: Support bayesianoptimization in v1alpha2

### PR DESCRIPTION
**Pull request by gaocegege**
Signed-off-by: Ce Gao <gaoce@caicloud.io>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/595)

### CODE REVIEW

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            experiment=experiment,
+            X_train=X_train,
+            y_train=y_train,
+            logger=self.logger,
+        )
+
+        lowerbound = np.array(algo_manager.lower_bound)
+        upperbound = np.array(algo_manager.upper_bound)
+        self.logger.debug("lowerbound: %r", lowerbound,
+                          extra={"StudyID": request.study_id})
```
Comment (johnugeorge) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: deprecated study_id

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        self.logger.debug("upperbound: %r", upperbound,
+                          extra={"StudyID": request.study_id})
```
Comment (johnugeorge) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: similar comment

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            max_features=service_params["max_features"],
+            model_type=service_params["model_type"],
+            logger=self.logger,
+        )
+        trials = []
+        x_next_list = alg.get_suggestion(request.request_number)
+        for x_next in x_next_list:
+            x_next = x_next.squeeze()
+            self.logger.debug("xnext: %r ", x_next, extra={
+                              "StudyID": request.study_id})
```
Comment (johnugeorge) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: Similar comment

File: pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+        self._X_train = self._mapping_params(X_train)
+        self.parse_X()
+
+        self._y_train = y_train
+        self._parse_metric()
+
+    @property
+    def experiment_name(self):
+        """ return the study id """
```
Comment (johnugeorge) [inline pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py]: s/study id/experiment name

File: pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        return self._experiment_name
+
+    @property
+    def experiment(self):
+        """ return the study configuration """
```
Comment (johnugeorge) [inline pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py]: Similar comment

File: pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    def X_train(self):
+        """ return the training data """
+        return self._X_train
+
+    @property
+    def y_train(self):
+        """ return the target of the training data"""
+        return self._y_train
+
+    def _parse_config(self):
```
Comment (johnugeorge) [inline pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py]: These utilities are already added in https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1alpha2/parsing_util.py  See https://github.com/kubeflow/katib/pull/435/files

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+from logging import getLogger, StreamHandler, INFO, DEBUG
+
+
+class BayesianService(api_pb2_grpc.SuggestionServicer):
+    def __init__(self, logger=None):
+        self.manager_addr = "katib-manager"
+        self.manager_port = 6789
+        if logger == None:
+            self.logger = getLogger(__name__)
+            FORMAT = '%(asctime)-15s StudyID %(studyid)s %(message)s'
```
Comment (andreyvelich) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: Should it be experimentname?

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        x_train = []
+        y_train = []
+        channel = grpc.beta.implementations.insecure_channel(
+            self.manager_addr, self.manager_port)
+        with api_pb2.beta_create_Manager_stub(channel) as client:
+            trialsrep = client.GetTrialList(api_pb2.GetTrialListRequest(
+                experiment_name=experiment_name
+            ))
+            for t in trialsrep.trials:
+                if t.status.condition == 2:
```
Comment (andreyvelich) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: Should we change 2 to 'api_pb2.TrialStatus.TrialConditionType.SUCCEEDED' ?

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                    gwfrep = client.GetObservationLog(
+                        api_pb2.GetObservationLogRequest(
+                            trial_name=t.name,
+                            metric_name=obj_name))
```
Comment (andreyvelich) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: You have missed timeout here. It should be like this: 
```
gwfrep = client.GetObservationLog(
                        api_pb2.GetObservationLogRequest(
                            trial_name=t.name,
                            metric_name=obj_name),10)
```

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+                    w = gwfrep.observation_log
+                    for ml in w.metrics_logs:
```
Comment (andreyvelich) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: `metrics_logs` -> `metric_logs`

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+                        if ml.name == obj_name:
+                            y_train.append(float(ml.values[-1].value))
```
Comment (andreyvelich) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: Should it be like this:
```
 if ml.metric.name == obj_name:
       y_train.append(float(ml.metric.value))
```
?

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+            ))
+            for t in trialsrep.trials:
+                if t.status.condition == 2:
+                    gwfrep = client.GetObservationLog(
+                        api_pb2.GetObservationLogRequest(
+                            trial_name=t.name,
+                            metric_name=obj_name))
+                    w = gwfrep.observation_log
+                    for ml in w.metrics_logs:
+                        if ml.name == obj_name:
```
Comment (andreyvelich) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: I don't think, that we need condition here, we select values only for `metric_name=obj_name`

**Comment by gaocegege** [comment]
Found some bugs during the test, I will fix it

/hold

File: pkg/suggestion/v1alpha2/bayesian_service.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+            for t in trialsrep.trials:
+                if t.status.condition == api_pb2.TrialStatus.TrialConditionType.SUCCEEDED:
+                    gwfrep = client.GetObservationLog(
+                        api_pb2.GetObservationLogRequest(
+                            trial_name=t.name,
+                            metric_name=obj_name), timeout)
+                    w = gwfrep.observation_log
+                    for ml in w.metric_logs:
+                        y_train.append(float(ml.metric.value))
+                        x_train.append(w.parameter_set)
```
Comment (andreyvelich) [inline pkg/suggestion/v1alpha2/bayesian_service.py]: I think, here it should be `t.spec.parameter_assignments.assignments` instead of `w.parameter_set`.

Comment (gaocegege) [reply pkg/suggestion/v1alpha2/bayesian_service.py]: Let me check.

**Comment by gaocegege** [comment]
@andreyvelich Fixed, thanks for the review.

File: pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
+""" module for algorithm manager """
+import numpy as np
+
+from pkg.api.v1alpha2.python import api_pb2
+
+from .utils import get_logger
+
+
+def deal_with_discrete(feasible_values, current_value):
```
Comment (johnugeorge) [inline pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py]: This contains utils that are already in parsing_utils. We can merge it later.

Comment (gaocegege) [reply pkg/suggestion/v1alpha2/bayesianoptimization/src/algorithm_manager.py]: SGTM.

**Comment by johnugeorge** [comment]
/lgtm
We should  also plan for unittests for suggestion algorithms

**Pull request merged**